import "mapreduce/yt/interface/protos/extension.proto";

package NYT.NTesting;

message TIntegral
{
    optional double DoubleField = 1;
    optional float FloatField = 2;
    optional int32 Int32Field = 3;
    optional int64 Int64Field = 4;
    optional uint32 Uint32Field = 5;
    optional uint64 Uint64Field = 6;
    optional sint32 Sint32Field = 7;
    optional sint64 Sint64Field = 8;
    optional fixed32 Fixed32Field = 9;
    optional fixed64 Fixed64Field = 10;
    optional sfixed32 Sfixed32Field = 11;
    optional sfixed64 Sfixed64Field = 12;
    optional bool BoolField = 13;
    enum TriBool
    {
        TRI_FALSE = 0;
        TRI_TRUE = 1;
        TRI_UNDEF = -1;
    }
    optional TriBool EnumField = 14;
}

message TRepeated
{
    repeated int32 Int32Field = 1;
}

message TRepeatedYtMode
{
    option (NYT.default_field_flags) = EXPERIMENTAL_SERIALIZATION_YT;
    repeated int32 Int32Field = 1;
}

message TWithTypeOptions
{
    enum Color
    {
        WHITE = 0;
        BLUE = 1;
        RED = -1;
    }

    message Embedded
    {
        optional Color ColorIntField = 1 [(NYT.flags) = ENUM_INT];
        optional Color ColorStringField = 2 [(NYT.flags) = ENUM_STRING];
        optional bytes AnyField = 3 [(NYT.flags) = ANY];
    }

    optional Color ColorIntField = 1 [(NYT.flags) = ENUM_INT];
    optional Color ColorStringField = 2 [(NYT.flags) = ENUM_STRING];
    optional bytes AnyField = 3 [(NYT.flags) = ANY];
    optional bytes OtherColumnsField = 4 [(NYT.flags) = OTHER_COLUMNS];
    optional Embedded EmbeddedField = 5 [(NYT.flags) = EXPERIMENTAL_SERIALIZATION_YT];
    repeated Color RepeatedEnumIntField = 6 [(NYT.flags) = EXPERIMENTAL_SERIALIZATION_YT, (NYT.flags) = ENUM_INT];
}

message TWithTypeOptions_TypeMismatch_EnumInt
{
    optional int64 EnumField = 1 [(NYT.flags) = ENUM_INT];
}

message TWithTypeOptions_TypeMismatch_EnumString
{
    optional string EnumField = 1 [(NYT.flags) = ENUM_STRING];
}

message TWithTypeOptions_TypeMismatch_Any
{
    optional string AnyField = 1 [(NYT.flags) = ANY];
}

message TWithTypeOptions_TypeMismatch_OtherColumns
{
    optional string OtherColumnsField = 1 [(NYT.flags) = OTHER_COLUMNS];
}

message TOneOf
{
    oneof Chooser
    {
        double DoubleField = 1;
        int32 Int32Field = 2;
    }
    optional bool BoolField = 3;
}

message TWithRequired
{
    required string RequiredField = 1;
    optional string NotRequiredField = 2;
};

message TAggregated
{
    optional string StringField = 1;
    optional bytes BytesField = 2;
    optional TIntegral NestedField = 3;
    optional TRepeated NestedRepeatedField = 4;
    optional TOneOf NestedOneOfField = 5;
    optional TAggregated NestedRecursiveField = 6;
}

message TAliased
{
    optional int32 Key = 1         [(NYT.key_column_name) = "key"];
    optional double Subkey = 2     [(NYT.key_column_name) = "subkey"];
    optional TAggregated Data = 3;
}

////////////////////////////////////////////////////////////////////////////////

message TUrlRow
{
    optional string Host = 1     [(NYT.column_name) = "Host"];
    optional string Path = 2     [(NYT.column_name) = "Path"];
    optional sint32 HttpCode = 3 [(NYT.column_name) = "HttpCode"];
}

message TRowFieldSerializationOption
{
    optional TUrlRow UrlRow_1 = 1 [(NYT.flags) = EXPERIMENTAL_SERIALIZATION_YT];
    optional TUrlRow UrlRow_2 = 2;
}

message TRowMessageSerializationOption
{
    option (NYT.default_field_flags) = EXPERIMENTAL_SERIALIZATION_YT;
    optional TUrlRow UrlRow_1 = 1;
    optional TUrlRow UrlRow_2 = 2;
}

message TRowMixedSerializationOptions
{
    option (NYT.default_field_flags) = EXPERIMENTAL_SERIALIZATION_YT;
    optional TUrlRow UrlRow_1 = 1;
    optional TUrlRow UrlRow_2 = 2 [(NYT.flags) = SERIALIZATION_PROTOBUF];
}

message TRowSerializedRepeatedFields
{
    option (NYT.default_field_flags) = EXPERIMENTAL_SERIALIZATION_YT;
    repeated int64 Ints = 1;
    repeated TUrlRow UrlRows = 2;
}

message TUrlRowWithColumnNames
{
    optional string Host = 1     [(NYT.column_name) = "Host_ColumnName", (NYT.key_column_name) = "Host_KeyColumnName"];
    optional string Path = 2     [(NYT.key_column_name) = "Path_KeyColumnName"];
    optional sint32 HttpCode = 3;
}

message TRowMixedSerializationOptions_ColumnNames
{
    option (NYT.default_field_flags) = EXPERIMENTAL_SERIALIZATION_YT;
    optional TUrlRowWithColumnNames UrlRow_1 = 1;
    optional TUrlRowWithColumnNames UrlRow_2 = 2 [(NYT.flags) = SERIALIZATION_PROTOBUF];
}
